src/Tracker.cpp

```diff
@@ -1895,3 +1968,157 @@ void Tracker::CollectNearestPoints(KeyFrame& kf, std::set<MapPoint*>& spNearestP
+
+    score_pair mp; mp.first = totalEntropyReduction; mp.second = mvKeyFrameBuffer.size() - 1;
+    vKeyframeScores.push_back(mp);
+
+    if(totalEntropyReduction > bestBufferKeyframeScore) //if we have a new winner
+    {
+        bestBufferKeyframeScore = totalEntropyReduction;
+        bestBufferKeyframeIndex = mvKeyFrameBuffer.size() - 1;
+    }
+
+
+}
+
+void Tracker::AddNewKeyFrameFromBuffer(int bufferPosition)
+{
+    if( (int)mvKeyFrameBuffer.size() < bufferPosition) //exceeding index
+        return;
+
+    static GVars3::gvar3<int> gvnCrossCamera("CrossCamera", 1, GVars3::HIDDEN|GVars3::SILENT);
+    //note: should check to make sure index doesnt exceed buffer size
+
+    //pull out the keyframe from the buffer
+    MultiKeyFrame* mpTempMKF =  mvKeyFrameBuffer[bufferPosition];
+
+    //set it's status to NOT a bufferkeyframe (it's getting inserted into the map!
+    mpTempMKF->isBufferMKF = false;
+
+    //iterate through all the measurements in the chosen MKF and all all the measurements to the map points
+    //boost::mutex::scoped_lock lock(mMap.mMutex);
+    for(KeyFramePtrMap::iterator it = mpTempMKF->mmpKeyFrames.begin(); it != mpTempMKF->mmpKeyFrames.end(); ++it)
+    {
+        std::string cName = (it->first);
+        KeyFrame* kf = (it->second);
+
+        //std::cout<<"At Keyframe %x " << &kf <<std::endl;
+
+        //iterate through each measurement map in the keyframe
+        std::vector<MapPoint*> ptsToDelete;
+        for(MeasPtrMap::iterator it_mp = kf->mmpMeasurements.begin(); it_mp != kf->mmpMeasurements.end(); ++it_mp)
+        {
+            MapPoint& kf_point = *(it_mp->first);
+
+            //IMPORTANT: check to see if the pointer is still in the map! It may have been trashed by the map maker
+            if(std::count(mMap.mlpPoints.begin(), mMap.mlpPoints.end(), &kf_point)) //is the point still in the map??
+            {
+                //if(kf_point.mMMData.spMeasurementKFs.count(kf)) //if its not in there already, or is bad (bad points taken car of when handed to map maker)
+                //continue;
+
+                //std::cout<<"inserting keyframe " << kf.mCamName << " at point %x " <<  &kf_point <<std::endl;
+                //boost::mutex::scoped_lock lock(kf->mMeasMutex);
+
+                kf_point.mMMData.spMeasurementKFs.insert(kf); //add measurements of the keyframe to the point
+            }
+            else //want to erase measurement to the point
+            {
+                //std::cout<<"erasing meas of nonexisting point..."<<std::endl;
+                //kf->EraseMeasurementOfPoint(&kf_point);
+                //kf_point.mMMData.spMeasurementKFs.erase(kf);
+
+                ptsToDelete.push_back(&kf_point);
+            }
+
+        }
+
+        for(int i=0; i<(int)ptsToDelete.size(); i++)
+        {
+            kf->EraseMeasurementOfPoint(ptsToDelete[i]);
+        }
+    }
+
+    mMapMaker.AddMultiKeyFrame(mpTempMKF);  // map maker takes ownership
```

Comment:
Shouldn't this copy it out of the buffer, so the buffer can be used for new keyframes? It seems you are leaving the retained keyframe in the buffer and deleting everything around it, but won't it get overwritten later?